### PR TITLE
[dynamo / DDP] - lazily compile submodules - to propagate real tensor strides to backend compiler

### LINKED ISSUE #113740: torch.compile segmentation fault
### 🐛 Describe the bug

Error stack:

```
BFD: DWARF error: section .debug_info is larger than its filesize! (0x5d4c26 vs 0x436168)
==== backtrace (tid:3341850) ====
 0 0x0000000000042520 __sigaction()  ???:0
 1 0x000000000016a659 _Py_XINCREF()  python-3.8.13/Include/object.h:531
 2 0x000000000016a659 _PyErr_SetObject()  python-3.8.13/Python/errors.c:115
 3 0x000000000016a659 _PyErr_SetString()  python-3.8.13/Python/errors.c:206
 4 0x000000000016a659 PyErr_SetString()  python-3.8.13/Python/errors.c:214
 5 0x0000000000001403 gpuAssert.constprop.0()  main.c:0
 6 0x0000000000001873 launch()  main.c:0
 7 0x000000000013bf90 cfunction_call_varargs()  python-3.8.13/Objects/call.c:758
 8 0x000000000013bf90 PyCFunction_Call()  python-3.8.13/Objects/call.c:773
 9 0x0000000000124631 _PyObject_MakeTpCall()  python-3.8.13/Objects/call.c:159
10 0x0000000000120481 _PyObject_Vectorcall()  python-3.8.13/Include/cpython/abstract.h:125
11 0x0000000000120481 call_function()  python-3.8.13/Python/ceval.c:4963
12 0x0000000000120481 _PyEval_EvalFrameDefault()  python-3.8.13/Python/ceval.c:3469
13 0x000000000011a461 PyEval_EvalFrameEx()  python-3.8.13/Python/ceval.c:741
14 0x000000000012c33c _PyFunction_Vectorcall()  python-3.8.13/Objects/call.c:436
15 0x000000000013e09c PyVectorcall_Call()  python-3.8.13/Objects/call.c:200
16 0x000000000013e09c PyObject_Call()  python-3.8.13/Objects/call.c:228
17 0x000000000011d870 do_call_core()  python-3.8.13/Python/ceval.c:5010
18 0x000000000011d870 _PyEval_EvalFrameDefault()  python-3.8.13/Python/ceval.c:3559
19 0x000000000011ab76 PyEval_EvalFrameEx()  python-3.8.13/Python/ceval.c:741
20 0x000000000011ab76 _PyEval_EvalCodeWithName()  python-3.8.13/Python/ceval.c:4298
21 0x000000000013b85c _PyFunction_Vectorcall()  python-3.8.13/Objects/call.c:436
22 0x000000000013b85c _Py_CheckFunctionResult()  python-3.8.13/Objects/call.c:25
23 0x000000000013b85c _PyObject_Vectorcall()  python-3.8.13/Include/cpython/abstract.h:128
24 0x000000000013b85c method_vectorcall()  python-3.8.13/Objects/classobject.c:60
25 0x000000000011c808 _PyObject_Vectorcall()  python-3.8.13/Include/cpython/abstract.h:127
26 0x000000000011c808 _Py_CheckFunctionResult()  python-3.8.13/Objects/call.c:25
27 0x000000000011c808 _PyObject_Vectorcall()  python-3.8.13/Include/cpython/abstract.h:128
28 0x000000000011c808 call_function()  python-3.8.13/Python/ceval.c:4963
29 0x000000000011c808 _PyEval_EvalFrameDefault()  python-3.8.13/Python/ceval.c:3515
30 0x000000000012c2a6 PyEval_EvalFrameEx()  python-3.8.13/Python/ceval.c:741
31 0x000000000012c2a6 _PyFunction_Vectorcall()  python-3.8.13/Objects/call.c:411
32 0x000000000011baab _PyObject_Vectorcall()  python-3.8.13/Include/cpython/abstract.h:127
33 0x000000000011baab _Py_CheckFunctionResult()  python-3.8.13/Objects/call.c:25
34 0x000000000011baab _PyObject_Vectorcall()  python-3.8.13/Include/cpython/abstract.h:128
35 0x000000000011baab call_function()  python-3.8.13/Python/ceval.c:4963
36 0x000000000011baab _PyEval_EvalFrameDefault()  python-3.8.13/Python/ceval.c:3500
37 0x000000000011ab76 PyEval_EvalFrameEx()  python-3.8.13/Python/ceval.c:741
38 0x000000000011ab76 _PyEval_EvalCodeWithName()  python-3.8.13/Python/ceval.c:4298
39 0x000000000012c33c _PyFunction_Vectorcall()  python-3.8.13/Objects/call.c:436
40 0x000000000011baab _PyObject_Vectorcall()  python-3.8.13/Include/cpython/abstract.h:127
41 0x000000000011baab _Py_CheckFunctionResult()  python-3.8.13/Objects/call.c:25
42 0x000000000011baab _PyObject_Vectorcall()  python-3.8.13/Include/cpython/abstract.h:128
43 0x000000000011baab call_function()  python-3.8.13/Python/ceval.c:4963
44 0x000000000011baab _PyEval_EvalFrameDefault()  python-3.8.13/Python/ceval.c:3500
45 0x000000000011a461 PyEval_EvalFrameEx()  python-3.8.13/Python/ceval.c:741
46 0x000000000012c33c _PyFunction_Vectorcall()  python-3.8.13/Objects/call.c:436
47 0x000000000011c808 _PyObject_Vectorcall()  python-3.8.13/Include/cpython/abstract.h:127
48 0x000000000011c808 _Py_CheckFunctionResult()  python-3.8.13/Objects/call.c:25
49 0x000000000011c808 _PyObject_Vectorcall()  python-3.8.13/Include/cpython/abstract.h:128
50 0x000000000011c808 call_function()  python-3.8.13/Python/ceval.c:4963
51 0x000000000011c808 _PyEval_EvalFrameDefault()  python-3.8.13/Python/ceval.c:3515
52 0x000000000011ab76 PyEval_EvalFrameEx()  python-3.8.13/Python/ceval.c:741
53 0x000000000011ab76 _PyEval_EvalCodeWithName()  python-3.8.13/Python/ceval.c:4298
54 0x000000000012c33c _PyFunction_Vectorcall()  python-3.8.13/Objects/call.c:436
55 0x000000000013e172 PyVectorcall_Call()  python-3.8.13/Objects/call.c:200
56 0x000000000013e172 _Py_CheckFunctionResult()  python-3.8.13/Objects/call.c:25
```

Python backtrace:

```
  File "<string>", line 6 in launcher
  File "torch/_inductor/triton_ops/autotune.py", line 190 in run
  File "/tmp/tmpz6xfo1g3/m7/cm7lt2eztbclmuwmkovykcboybniz3mtoovrrpp6qu7kem6q2cpx.py", line 11267 in call
  File "torch/_inductor/compile_fx.py", line 248 in run
  File "torch/_functorch/aot_autograd.py", line 1249 in call_func_with_args
  File "torch/_functorch/aot_autograd.py", line 2168 in forward
  File "torch/autograd/function.py", line 506 in apply
  File "torch/_functorch/aot_autograd.py", line 1224 in g
  File "torch/_functorch/aot_autograd.py", line 1249 in call_func_with_args
  File "torch/_functorch/aot_autograd.py", line 1900 in runtime_wrapper
  File "torch/_functorch/aot_autograd.py", line 2403 in debug_compiled_function
  File "torch/_functorch/aot_autograd.py", line 1224 in g
  File "torch/_functorch/aot_autograd.py", line 2836 in forward
  File "torch/_dynamo/eval_frame.py", line 209 in _fn
  File "torch/_dynamo/backends/distributed.py", line 247 in forward
  File "torch/nn/modules/module.py", line 1501 in _call_impl
  File "<eval_with_key>.253", line 347 in forward
  File "torch/_dynamo/eval_frame.py", line 209 in _fn
  File "torch/_dynamo/eval_frame.py", line 82 in forward
  File "torch/nn/modules/module.py", line 1501 in _call_impl
  File "foo.py", line 710 in forward
  File "torch/nn/modules/module.py", line 1501 in _call_impl
  File "torch/nn/parallel/distributed.py", line 1110 in _run_ddp_forward
  File "torch/nn/parallel/distributed.py", line 1156 in forward
  File "torch/_dynamo/eval_frame.py", line 209 in _fn
  File "torch/_dynamo/eval_frame.py", line 82 in forward
  File "torch/nn/modules/module.py", line 1501 in _call_impl
```

Running `$python /tmp/tmpz6xfo1g3/m7/cm7lt2eztbclmuwmkovykcboybniz3mtoovrrpp6qu7kem6q2cpx.py` also fails with a different error:

```
    assert_size_stride(buf1711, (10, 480, 1, 30), (14400, 1, 14400, 480))
AssertionError: expected size 480==480, stride 30==1 at dim=1
```

### Error logs

_No response_

### Minified repro

_No response_

### Versions

PyTorch 2.0.1

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

### LINKED ISSUE #113812: DDPOptimizer and Inductor output stride mismatch issue
### 🐛 Describe the bug

When `torch._inductor.config.keep_output_stride=True` (default value), we do preserve the strides of end-user-visible Dynamo graph output.

However, DDPOptimizer then splits the Dynamo graph into smaller subgraphs, and calls Inductor compile on each of the subgraphs. DDPOptimizer's design expects the ‘output’ of those smaller subgraphs to have their strides be preserved as well, but Inductor doesn't have a contract to preserve them, hence causing stride mismatch error e.g. in https://github.com/pytorch/pytorch/issues/113740, https://github.com/pytorch/pytorch/issues/107362 and also OSS dist benchmarks.

One solution (suggested by Elias) is to let DDPOptimizer not do graph-splitting and compilation until during the first execution with real tensors, so that we can guarantee that the actual strides produced by Inductor-compiled subgraph output is propagated downstream.

We should also have a unit test that repros the issue and verify that this solution fixes it.

Related issue: https://github.com/pytorch/pytorch/issues/107362#issuecomment-1684434426

cc. @wconstab @eellison 

### Versions

-

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/113812, https://github.com/pytorch/pytorch/issues/102591, Probably fixes: https://github.com/pytorch/pytorch/issues/113740, https://github.com/pytorch/pytorch/issues/113786, https://github.com/pytorch/pytorch/issues/113788

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @aakhundov @kadeng @kiukchung @lucasllc @d4l3k

### CODE REVIEW

File: torch/_dynamo/backends/distributed.py
Comment (jon-chuang) [inline torch/_dynamo/backends/distributed.py:463]: Logging the final graph below doesn't seem so meaningful anymore. Maybe only log it when the entire graph module is run?

File: torch/_dynamo/backends/distributed.py
Comment (jon-chuang) [inline torch/_dynamo/backends/distributed.py:427]: Add this back please

File: torch/_dynamo/backends/distributed.py
Comment (jon-chuang) [inline torch/_dynamo/backends/distributed.py:463]: Do we really need to keep this? We no longer actually propagate any fake_args information.

File: torch/_dynamo/backends/distributed.py
Comment (jon-chuang) [inline torch/_dynamo/backends/distributed.py:430]: You can remove this too

File: torch/_dynamo/backends/distributed.py
Comment (jon-chuang) [inline torch/_dynamo/backends/distributed.py:434]: Remove this comment

File: torch/_dynamo/backends/distributed.py
Comment (jon-chuang) [inline torch/_dynamo/backends/distributed.py:421]: Please update comments here to reflect lazy compilation

**Comment by jon-chuang** [comment]
I think we can remove this restriction with this PR, right? @eellison @yf225 

> torch._inductor.config.keep_output_stride=True

**Comment by ezyang** [comment]
Wow! If this works this is great, a lot simpler code

File: torch/_dynamo/backends/distributed.py
Comment (ezyang) [inline torch/_dynamo/backends/distributed.py:338]: Only populate fake args if not compiled

File: torch/_dynamo/backends/distributed.py
Comment (jon-chuang) [inline torch/_dynamo/backends/distributed.py:341]: Maybe we also want to handle fakeified? 

not `torch._subclasses.fake_tensor.is_fake(..)`

**Comment by ezyang** [comment]
@wconstab for the final call, but this looks pretty promising

File: torch/_dynamo/backends/distributed.py
Comment (wconstab) [inline torch/_dynamo/backends/distributed.py:350]: is it actually necessary to do the first trace with fake args?

IIRC it may have been necessary at one point, but i was trying something similar recently and it seemed like maybe this requirement is relaxed now?

Comment (jon-chuang) [reply torch/_dynamo/backends/distributed.py:350]: If you run with real args, it will run the real computation, which may include mutations.

File: torch/_dynamo/backends/distributed.py
Comment (wconstab) [inline torch/_dynamo/backends/distributed.py:451]: how come this is removed?

Comment (jon-chuang) [reply torch/_dynamo/backends/distributed.py:451]: We don't need the example outputs when running `SubmodCompiler` anymore.

We just need to replace the submodules with lazily-compiling submodules.

File: torch/_dynamo/backends/distributed.py
Comment (wconstab) [inline torch/_dynamo/backends/distributed.py:326]: i think we should rename/cleanup this wrapper class a bit more.

In my other PR (https://github.com/pytorch/pytorch/pull/114186) i renamed it to 'UnwrapperModule' bc it seems like that's more accurate.  And i updated the comments to explain more why i think we need the unwrapping to happen.

In this case maybe just call it a LazyCompileModule or something instead, but still good to take some of my updated comments.

Comment (jon-chuang) [reply torch/_dynamo/backends/distributed.py:326]: Sounds good.

File: torch/_dynamo/backends/distributed.py
Comment (wconstab) [reply torch/_dynamo/backends/distributed.py:451]: oh i see. we no longer care about propagating 'args' during this traversal, so i guess we are propagating 'None' value implicitly instead of propagating real or fake values like we used to?

If so, please add a comment. (but i think that's fine).

File: torch/_dynamo/backends/distributed.py
Comment (wconstab) [inline torch/_dynamo/backends/distributed.py:351]: maybe simplify to just `self.submod = self.compiler(self.submod...)`?

File: torch/_dynamo/backends/distributed.py
Comment (jon-chuang) [reply torch/_dynamo/backends/distributed.py:451]: Yes let me double check if the behaviour when returning None in the `run_node` is to fallback on that node's hard-coded args. Maybe the kwargs checking needs to be deferred to the lazy submodule if not.

File: test/distributed/test_dynamo_distributed.py
Comment (wconstab) [inline test/distributed/test_dynamo_distributed.py:595]: did you confirm that this model triggered the expected graph-splitting and exposed the strides bug?

File: test/distributed/test_dynamo_distributed.py
Comment (wconstab) [inline test/distributed/test_dynamo_distributed.py:572]: is it really necessary to assert this after we patch it above?

File: torch/_dynamo/backends/distributed.py
Comment (wconstab) [reply torch/_dynamo/backends/distributed.py:350]: yea, you're right.

**Review by wconstab — APPROVED** [review]
Thanks for fixing this, it has been a real pain.  

Please address the nit comments but otherwise LGTM.

File: torch/_dynamo/backends/distributed.py
Comment (jon-chuang) [reply torch/_dynamo/backends/distributed.py:351]: tried that, it says cant set already existing mod. So we need to del first.

File: test/distributed/test_dynamo_distributed.py
Comment (jon-chuang) [reply test/distributed/test_dynamo_distributed.py:595]: Yes, as per comment 
https://github.com/pytorch/pytorch/pull/114154#issuecomment-1819698382

When removing this assert https://github.com/pytorch/pytorch/pull/114154#issuecomment-1819673192, it fails with mismatched strides error

File: test/distributed/test_dynamo_distributed.py
Comment (jon-chuang) [reply test/distributed/test_dynamo_distributed.py:572]: I had some issue with patch.object on config previously. Maybe I'll convert to `@config.patch(..)` instead since it is also supported.

Comment (jon-chuang) [reply test/distributed/test_dynamo_distributed.py:572]: Anw, I think the assert is not a bad thing

File: torch/_dynamo/backends/distributed.py
Comment (jon-chuang) [reply torch/_dynamo/backends/distributed.py:451]: Actually, kwargs checking does not need to be deferred. If there are any kwargs, we will still error, since even if the kwargs value are `None`, the node would still have been passed a kwarg, which we do not currently support.

**Comment by DanilBaibak** [comment]
Hi @jon-chuang! Sorry for reverting your change but the lower PR in the stack #113396 is failing a bunch of internal build jobs. We will let you know when we fix the base PR.

**Comment by clee2000** [comment]
@wconstab Could you take a look at https://www.internalfb.com/diff/D51644971?dst_version_fbid=337126938929132&transaction_fbid=1486960798528659 and help get this remerged?

**Comment by jon-chuang** [comment]
@clee2000 , if the logs could be shared (esp if just error message is informative enough) I can try to fix too.

**Comment by jon-chuang** [comment]
@wconstab perhaps this may be my initial fears coming to be realized - compiling under a different set of conditions to when it is triggered directly by Dynamo seems to introduce some errors. Anw, I hope we can fix it.

**Comment by jon-chuang** [comment]
You could also test https://github.com/pytorch/pytorch/pull/114779 where I added the commits cleanly.

**Comment by clee2000** [comment]
Sorry for the late reply, I think this is the logs?
```
<path>/torch/_inductor/compile_fx.py", line 611, in run
    return model(new_inputs)
  File "<other path>/torch/_inductor/codecache.py", line 885, in _run_from_cache
    return compiled_graph.compiled_artifact(inputs)
  File "<tmp path>/cft4ryxjjtuvcjdt7djh6bmiarju3btzxkaryuyydba5nvxgconz.py", line 215, in call
    assert_size_stride(tangents_1, (s4, 1024 + s16), (1024 + s16, 1))
NameError: name 's16' is not defined
```

I'll import the other PR, but I'm not sure if the failing test will actually get run on it

**Comment by jon-chuang** [comment]
> NameError: name 's16' is not defined

This can come from anywhere honestly, hard to debug without an exact repro

**Comment by eellison** [comment]
cc @wconstab - maybe you could debug the internal issue ?

**Comment by wconstab** [comment]
@yf225 is looking into it already, any ideas yet?

**Comment by yf225** [comment]
> @yf225 is looking into it already, any ideas yet?

I can repro the issue, still looking at it

File: torch/_dynamo/backends/distributed.py
Comment (yf225) [inline torch/_dynamo/backends/distributed.py:356]:
```suggestion
new_submod = self.compiler(self.submod, args)
```
@jon-chuang If we make this change to use real tensors to do the compile, the internal test will also pass.

Comment (jon-chuang) [reply torch/_dynamo/backends/distributed.py:356]: Hmm I see. I just want to double check that we will not run the real computation on the real args. I think it is fine as we normally already pass real example inputs to our compilers.

Comment (jon-chuang) [reply torch/_dynamo/backends/distributed.py:356]: I guess there's no simple repro to add to the testing path? It would probably be better if we can find one, but I get that it's probably also hard to reproduce.

Comment (jon-chuang) [reply torch/_dynamo/backends/distributed.py]: @yf225 would you mind importing this PR (including new commits) into the meta CI? Just to confirm everything is solid before we merge again

Comment (yf225) [reply torch/_dynamo/backends/distributed.py]: @jon-chuang I imported and tested it internally, and can confirm that everything works. We are ready to merge this PR.

**Review by yf225 — APPROVED** [review]
Thanks @jon-chuang !

**Pull request closed without merging**
